src/helper.h

```diff
@@ -140,23 +134,25 @@ class helper
 
 class statefulhelper : public helper
 {
-    CBDATA_CLASS(statefulhelper);
-
 public:
+    using Pointer = RefCount<statefulhelper>;
     typedef std::unordered_map<Helper::ReservationId, helper_stateful_server *> Reservations;
 
-    inline statefulhelper(const char *name) : helper(name) {}
     inline ~statefulhelper() {}
 
 public:
+    static Pointer Make(const char *name);
+
```

Comment:
Please revert the code style to using C++ construction instead of a maker-factory style.

```suggestion
explicit statefulhelper(const char *name) : helper(name) {}
```

Reply:
> > Amos: Please revert the code style to using C++ construction instead of a maker-factory style.
> 
> Why?! PR code (returning a smart pointer) is safer than raw `new` (returning a pointer that can easily leak).

All callers are immediately and unconditionally de-referencing the pointer - regardless of whether it is raw or smart. That invalidates the claim of their being any difference in safety.

> The new API also directs new helper users to proper pointer storage. As an added bonus, this `X::Make()` upgrade significantly reduces a chance that we have missed a raw X allocation somewhere in this PR or will miss it somewhere in backporting.


> Overall, it is an improvement that should not be reverted IMO.

IMO simplicity is better API design. The result of `new` should be stored in a `Pointer` - that is a constant for all types in Squid with `Pointer` defined. This is gratuitously forcing developers to work with an unusual (for C++) allocation mechanism.

> 
> > Eduard: the factory style would involve some dynamic type selection
> 
> Let's ignore unimportant terminology differences. The essence of this change request is clear -- replace `X::Make()` with `new X()`. I see no justification for going back to raw pointers, but perhaps I am missing something. In either case, we can save time by _not_ arguing about the definition of "maker-factory style".

Nod. This is not a request to "go back to raw pointers" it is a request to remove a layer of abstraction between the construction and Pointer storage.